### ISSUE TITLE
Headless --import --quit crashes editor

### ISSUE THREAD

**Issue by TCROC** [godotengine/godot#98062]
### Tested versions

Reproduceable in 4.4.dev [842f98239713fd10cfd648cd6aa3781895f289eb]

### System information

Linux Pop OS 22.04

### Issue description

When importing with --headless --quit args, the godot editor will crash.  This appears to be due to unload_editor_addons() method that was introduced to being called in ``Main::iteration`` here:

https://github.com/godotengine/godot/commit/c151c7dce0b3802ddf2caa2bb2d6a227d7326c3f

The solution is simple.  Simply clear the ``addon_name_to_plugin`` so that we don't try to memdelete things that have already had memdelete called on them.  There is no need to keep track of these plugins in a list after they have been deleted from memory (at least so far as I can tell).  I did this in my fork over here and will be putting in a PR for it: https://github.com/Lange-Studios/godot/commit/96ab05e89ff7d4f9f2865b04d3b7bf2b489b009f

### Steps to reproduce

I suspect this may be a more challenging one to reproduce.  Likely it will not occur if the editor exits after the first time ``unload_editor_addons`` is called.  I suspect this to be a race condition that I somehow introduced in my project.  But if you can get it into this state, you should get segfault crashes due to ``remove_editor_plugin`` being called on plugins that have already had ``memdelete`` called.

### Minimal reproduction project (MRP)

N/A

**Comment by jamie-pate** [comment]
`ProgressDialog::_update_ui(...)` calls Main::iteration() without checking if it returns `true` to exit

```
EditorNode::unload_editor_addons(EditorNode * const this) (/home/jpate/build/src/godot/editor/editor_node.cpp:3403)
Main::iteration() (/home/jpate/build/src/godot/main/main.cpp:4265)
ProgressDialog::_update_ui(ProgressDialog * const this) (/home/jpate/build/src/godot/editor/progress_dialog.cpp:133)
ProgressDialog::add_task(ProgressDialog * const this, const String & p_task, const String & p_label, int p_steps, bool p_can_cancel) (/home/jpate/build/src/godot/editor/progress_dialog.cpp:199)
EditorNode::progress_add_task(const String & p_task, const String & p_label, int p_steps, bool p_can_cancel) (/home/jpate/build/src/godot/editor/editor_node.cpp:4936)
EditorProgress::EditorProgress(EditorProgress * const this, const String & p_task, const String & p_label, int p_amount, bool p_can_cancel) (/home/jpate/build/src/godot/editor/editor_node.cpp:188)
EditorFileSystem::reimport_files(EditorFileSystem * const this, const Vector<String> & p_files) (/home/jpate/build/src/godot/editor/editor_file_system.cpp:2657)
EditorFileSystem::_update_scan_actions(EditorFileSystem * const this) (/home/jpate/build/src/godot/editor/editor_file_system.cpp:804)
EditorFileSystem::_notification(EditorFileSystem * const this, int p_what) (/home/jpate/build/src/godot/editor/editor_file_system.cpp:1494)
EditorFileSystem::_notificationv(EditorFileSystem * const this, int p_notification, bool p_reversed) (/home/jpate/build/src/godot/editor/editor_file_system.h:141)
Object::notification(Object * const this, int p_notification, bool p_reversed) (/home/jpate/build/src/godot/core/object/object.cpp:873)
SceneTree::_process_group(SceneTree * const this, SceneTree::ProcessGroup * p_group, bool p_physics) (/home/jpate/build/src/godot/scene/main/scene_tree.cpp:962)
SceneTree::_process(SceneTree * const this, bool p_physics) (/home/jpate/build/src/godot/scene/main/scene_tree.cpp:1039)
SceneTree::process(SceneTree * const this, double p_time) (/home/jpate/build/src/godot/scene/main/scene_tree.cpp:526)
Main::iteration() (/home/jpate/build/src/godot/main/main.cpp:4154)
OS_LinuxBSD::run(OS_LinuxBSD * const this) (/home/jpate/build/src/godot/platform/linuxbsd/os_linuxbsd.cpp:962)
main(int argc, char ** argv) (/home/jpate/build/src/godot/platform/linuxbsd/godot_linuxbsd.cpp:85)
```

**Comment by TCROC** [comment]
> `ProgressDialog::_update_ui(...)` calls Main::iteration() without checking if it returns `true` to exit
> 
> 
> 
> This happens if you use `--quit` without the `--import` flag also.. (--import doesn't need --quit)
> 
> 
> 
> My CI import workflow:
> 
> ```Bash
> 
> set -e
> 
> # run the import process
> 
> godot --headless -e --import
> 
> # Run godot editor and quit immediately
> 
> godot --headless -e --quit
> 
> ```
> 
> 
> 
> This forces it to run once to import, and a 2nd time which can exit with any errors caused by resources that imported incorrectly
> 
> 
> 
> ```
> 
> EditorNode::unload_editor_addons(EditorNode * const this) (/home/jpate/build/src/godot/editor/editor_node.cpp:3403)
> 
> Main::iteration() (/home/jpate/build/src/godot/main/main.cpp:4265)
> 
> ProgressDialog::_update_ui(ProgressDialog * const this) (/home/jpate/build/src/godot/editor/progress_dialog.cpp:133)
> 
> ProgressDialog::add_task(ProgressDialog * const this, const String & p_task, const String & p_label, int p_steps, bool p_can_cancel) (/home/jpate/build/src/godot/editor/progress_dialog.cpp:199)
> 
> EditorNode::progress_add_task(const String & p_task, const String & p_label, int p_steps, bool p_can_cancel) (/home/jpate/build/src/godot/editor/editor_node.cpp:4936)
> 
> EditorProgress::EditorProgress(EditorProgress * const this, const String & p_task, const String & p_label, int p_amount, bool p_can_cancel) (/home/jpate/build/src/godot/editor/editor_node.cpp:188)
> 
> EditorFileSystem::reimport_files(EditorFileSystem * const this, const Vector<String> & p_files) (/home/jpate/build/src/godot/editor/editor_file_system.cpp:2657)
> 
> EditorFileSystem::_update_scan_actions(EditorFileSystem * const this) (/home/jpate/build/src/godot/editor/editor_file_system.cpp:804)
> 
> EditorFileSystem::_notification(EditorFileSystem * const this, int p_what) (/home/jpate/build/src/godot/editor/editor_file_system.cpp:1494)
> 
> EditorFileSystem::_notificationv(EditorFileSystem * const this, int p_notification, bool p_reversed) (/home/jpate/build/src/godot/editor/editor_file_system.h:141)
> 
> Object::notification(Object * const this, int p_notification, bool p_reversed) (/home/jpate/build/src/godot/core/object/object.cpp:873)
> 
> SceneTree::_process_group(SceneTree * const this, SceneTree::ProcessGroup * p_group, bool p_physics) (/home/jpate/build/src/godot/scene/main/scene_tree.cpp:962)
> 
> SceneTree::_process(SceneTree * const this, bool p_physics) (/home/jpate/build/src/godot/scene/main/scene_tree.cpp:1039)
> 
> SceneTree::process(SceneTree * const this, double p_time) (/home/jpate/build/src/godot/scene/main/scene_tree.cpp:526)
> 
> Main::iteration() (/home/jpate/build/src/godot/main/main.cpp:4154)
> 
> OS_LinuxBSD::run(OS_LinuxBSD * const this) (/home/jpate/build/src/godot/platform/linuxbsd/os_linuxbsd.cpp:962)
> 
> main(int argc, char ** argv) (/home/jpate/build/src/godot/platform/linuxbsd/godot_linuxbsd.cpp:85)
> 
> ```

I have a PR here with a fix: https://github.com/godotengine/godot/pull/98063

Technically the actual issue that was causing the crash was not clearing the list after memdeleting all of the elements. The PR simply clears that list after all elements are deleted.

**Comment by jamie-pate** [comment]
> Technically the actual issue that was causing the crash was not clearing the list after memdeleting all of the elements. The PR simply clears that list after all elements are deleted.

The original author probably expected to not have to worry about that since they thought returning true means exiting the program, but I agree that clearing the data structure after deleting the memory it's pointing to pretty good practice..

**Comment by TCROC** [comment]
Makes sense! Thanks for the review! :)

**Comment by jamie-pate** [comment]
Workaround: pass `--quit-after 10` instead of `--quit` to delay the quit by 10s ... hopefully it's long enough to avoid the crash in your CI

**Comment by TCROC** [comment]
> Workaround: pass `--quit-after 10` instead of `--quit` to delay the quit by 10s ... hopefully it's long enough to avoid the crash in your CI
> 
> 

Thank you! I have my own fork with the fix, but this may come in handy in the future when I run into things! :)

**Comment by jamie-pate** [comment]
> > Technically the actual issue that was causing the crash was not clearing the list after memdeleting all of the elements. The PR simply clears that list after all elements are deleted.
> 
> The original author probably expected to not have to worry about that since they thought returning true means exiting the program, but I agree that clearing the data structure after deleting the memory it's pointing to pretty good practice..

Upon further inspection it looks like this:

```c++
EditorNode::unload_editor_addons(...)
Main::iteration() //B
ProgressDialog::_update_ui(...) // doesn't handle exit code, gets called many times before A resumes
...
EditorFileSystem::reimport_files(...)
EditorFileSystem::_update_scan_actions()
...
SceneTree::process(SceneTree * const this, double p_time) (/home/jpate/build/src/godot/scene/main/scene_tree.cpp:526)
Main::iteration() //A
OS_LinuxBSD::run() // handles exit code
```

* The iteration `A` ends up calling iteration `B` many times while scanning and reimporting files.. 
* The plugins are deleted immediately the first time `B` is called
* Every subsequent B call of `iteration()` will have no plugins and won't be able to import any resources supplied by the plugins (use --import instead if that's what you want)
* Eventually `_update_scan_actions()` will end and we continue with iteration() `A`, and `run()` will read the return value and exit the program.

It would be _more correct_ to add a parameter to `iteration()` that lets it know it is being called outside of `run()` and therefore the return value is ignored... In that case we could avoid any cleanup logic in the `B` iteration...
That would be a pretty fundamental change and while the current logic is still a bit more complicated than I'd like, it ends up working correctly.

**Comment by TCROC** [comment]
> > > Technically the actual issue that was causing the crash was not clearing the list after memdeleting all of the elements. The PR simply clears that list after all elements are deleted.
> > 
> > 
> > The original author probably expected to not have to worry about that since they thought returning true means exiting the program, but I agree that clearing the data structure after deleting the memory it's pointing to pretty good practice..
> 
> Upon further inspection it looks like this:
> 
> ```c++
> EditorNode::unload_editor_addons(...)
> Main::iteration() //B
> ProgressDialog::_update_ui(...) // doesn't handle exit code, gets called many times before A resumes
> ...
> EditorFileSystem::reimport_files(...)
> EditorFileSystem::_update_scan_actions()
> ...
> SceneTree::process(SceneTree * const this, double p_time) (/home/jpate/build/src/godot/scene/main/scene_tree.cpp:526)
> Main::iteration() //A
> OS_LinuxBSD::run() // handles exit code
> ```
> 
>     * The iteration `A` ends up calling iteration `B` many times while scanning and reimporting files..
> 
>     * The plugins are deleted immediately the first time `B` is called
> 
>     * Every subsequent B call of `iteration()` will have no plugins and won't be able to import any resources supplied by the plugins (use --import instead if that's what you want)
> 
>     * Eventually `_update_scan_actions()` will end and we continue with iteration() `A`, and `run()` will read the return value and exit the program.
> 
>     * It all happens on the main thread
> 
> 
> It would be _more correct_ to add a parameter to `iteration()` that lets it know it is being called outside of `run()` and therefore the return value is ignored... In that case we could avoid any cleanup logic in the `B` iteration... That would be a pretty fundamental change and while the current logic is still a bit more complicated than I'd like, it ends up working correctly.

This is outside of my familiarity personally.  I think the existing PR is a good PR since its good practice to clean up data structures anyways.  BUT, if what you suggest will increase stability, I would love to see an additional PR that puts this into effect :).  I just don't think I'm the right person to do that PR.

TL; DR;

I think the existing PR is still applicable even with this suggestion.  But I am definitely supportive of increased soundness in logic and improved stability! :)
